src/ann_solo/reader.py

```diff
@@ -231,6 +299,324 @@ def get_version(self) -> str:
+                                    'r' if mode == 'r' else 'w')
+
+    def close_store(self) -> None:
+        """
+        Close the hdf5 spectral library store after read/write is done.
+        """
+        if self.hdf5_store is not None:
+            self.hdf5_store.close()
+            self.hdf5_store = None
+
+    def get_all_spectra_ids(self) -> Iterator[str]:
+        """
+        Close the hdf5 spectral library store after write/read is done.
+
+        Returns
+        -------
+        Iterator[str]
+            An iterator of the keys/ids of spectra in the library store.
+
+        """
+        for spec_id in self.hdf5_store.keys():
+            yield spec_id
+
+    def write_spectrum_to_library(self, spectrum : MsmsSpectrum) -> None:
+        """
+        Gets an Msmsspectrum object, reads attributes, and stores it in the
+        spectral library store for future retrieval.
+
+        Parameters
+        ----------
+        spectrum : MsmsSpectrum
+            The MsmsSpectrum object.
+
+        """
+        # Create a new group under the same key
+        group = self.hdf5_store.create_group(f'{spectrum.identifier}')
+
+        group.create_dataset('peptide', data=spectrum.peptide)
+        group.create_dataset('precursor_charge',
+                             data=spectrum.precursor_charge)
+        group.create_dataset('precursor_mz', data=spectrum.precursor_mz)
+        group.create_dataset('is_decoy', data=spectrum.is_decoy)
+        group.create_dataset('mz', data=spectrum.mz)
+        group.create_dataset('intensity', data=spectrum.intensity)
+        # Encode annotation
+        annotation = np.array([[annotation.charge,
+                              annotation.calc_mz,
+                              _encode_ion_type(annotation.ion_type),
+                              annotation.ion_index]
+                             if annotation is not None
+                             else [0, 0, 0, 0]
+                             for annotation in spectrum.annotation])
+
+        group.create_dataset('annotation', data=annotation)
+
+        self.hdf5_store.flush()
+
+    def read_spectrum_from_library(self, spec_id : str)-> MsmsSpectrum:
+        """
+        Gets a library spectrum id and returns the corresponding spectrum as
+        an Msmsspectrum object.
+
+        Parameters
+        ----------
+        spec_id : string
+            A library spectrum id.
+
+        Returns
+        -------
+        MsmsSpectrum
+            An MsmsSpectrum object.
+        """
+        spectrum_specs = self.hdf5_store[f'{spec_id}']
+
+        # Decode annotation
+        annotation = [PeptideFragmentAnnotation(annotation[0].astype(np.int),
+                                                annotation[1],
+                                                _decode_ion_type(annotation[2]),
+                                                annotation[3].astype(np.int))
+                       if not np.all(annotation==0)
```

Comment:
**issue:** I don't think this is correct, use [`numpy.count_nonzero`](https://numpy.org/doc/stable/reference/generated/numpy.count_nonzero.html) instead.